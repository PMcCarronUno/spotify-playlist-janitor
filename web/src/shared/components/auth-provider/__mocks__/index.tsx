import React from 'react';
export const AuthProvider = ({ children }) => <div>{children}</div>;
