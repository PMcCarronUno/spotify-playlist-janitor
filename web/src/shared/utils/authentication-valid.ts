export const isAuthDataValid = (token?: string, expiresOn?: number) => {
  const now = Math.floor(new Date(Date.now()).getTime() / 1000);

  return (
    typeof token === "string" &&
    typeof expiresOn === "number" &&
    now < expiresOn
  );
};
