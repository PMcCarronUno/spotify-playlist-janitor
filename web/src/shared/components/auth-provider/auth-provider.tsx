import { memo, PropsWithChildren } from "react";
import { Navigate } from "react-router-dom";

import { useUser } from "shared/state/user";

import { isAuthDataValid } from "shared/utils/authentication-valid";

type AuthProviderProps = {
  shouldBeAuthorised: boolean;
  redirectPath: string;
};

export const AuthProvider = memo(
  ({
    shouldBeAuthorised,
    redirectPath,
    children,
  }: PropsWithChildren<AuthProviderProps>) => {
    const user = useUser();
    const loggedIn = isAuthDataValid(user?.access_token, user?.expires_on);

    if (loggedIn === shouldBeAuthorised) {
      return <>{children}</>;
    }

    return (
      <Navigate
        data-testid={"redirect-to-" + redirectPath.replace("/", "")}
        to={redirectPath}
      />
    );
  }
);

AuthProvider.displayName = "AuthProvider";
