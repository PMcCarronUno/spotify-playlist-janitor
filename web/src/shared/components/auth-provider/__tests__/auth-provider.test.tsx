import React from "react";
import { render } from "@testing-library/react";

import { isAuthDataValid } from "shared/utils/authentication-valid";
import { AuthProvider } from "../auth-provider";

jest.mock("shared/state/user");
jest.mock("shared/utils/authentication-valid");

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  Navigate: jest.fn(() => <div>Redirect Page</div>),
}));

describe("AuthProvider", () => {
  let container: HTMLElement;

  beforeEach(() => {
    jest.mocked(isAuthDataValid).mockImplementation(() => true);

    ({ container } = render(
      <AuthProvider>
        <div>Child Page</div>
      </AuthProvider>
    ));
  });

  it("should render a the children props if auth data is valid", () => {
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render a redirect component if auth data is invalid", () => {
    jest.mocked(isAuthDataValid).mockImplementation(() => false);

    ({ container } = render(
      <AuthProvider>
        <div>Child Page</div>
      </AuthProvider>
    ));

    expect(container.firstChild).toMatchSnapshot();
  });
});
