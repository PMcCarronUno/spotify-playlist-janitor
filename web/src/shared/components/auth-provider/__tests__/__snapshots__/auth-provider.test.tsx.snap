// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthProvider should render a redirect component if auth data is invalid 1`] = `
<div>
  Redirect Page
</div>
`;

exports[`AuthProvider should render a the children props if auth data is valid 1`] = `
<div>
  Child Page
</div>
`;
